File: backend/learning/sharshabeel.py
"""شرشبيل — Barwaaqo Skills assistant (rule-based + optional OpenAI)."""
import os
import re

from courses.models import Course


BOT_NAME = "شرشبيل"

FAQ_ENTRIES = [
    {
        "keys": ["experience", "prior experience", "beginner", "start from zero"],
        "answer": (
            "No prior experience is needed. Many courses are beginner-friendly and take you step by step "
            "from basics to job-ready skills."
        ),
    },
    {
        "keys": ["xp", "points", "streak"],
        "answer": (
            "You earn 10 XP per lesson completed and 100 XP when you finish a course. "
            "Your XP and streak appear on your dashboard."
        ),
    },
    {
        "keys": ["pace", "own time", "schedule", "self-paced"],
        "answer": "All courses are self-paced — study whenever it fits your schedule.",
    },
    {
        "keys": ["free", "cost", "price", "pay"],
        "answer": (
            "All courses are free. Create an account, verify your email, and enroll in up to 3 courses at a time."
        ),
    },
    {
        "keys": ["enroll", "how many", "3 course", "slot", "enrollment limit"],
        "answer": (
            "You can actively enroll in up to 3 courses. Finish a course or unenroll from My Courses to free a slot."
        ),
    },
    {
        "keys": ["certificate", "cert"],
        "answer": (
            "Complete all lessons, pass the final quiz, and you earn a certificate you can verify on the site."
        ),
    },
    {
        "keys": ["verify email", "email verification", "verification"],
        "answer": (
            "After signup, check your email for a verification link. You must verify before enrolling in courses."
        ),
    },
    {
        "keys": ["quiz", "assessment", "pass"],
        "answer": "Each course has a final quiz. You need 70% or higher to pass and unlock your certificate.",
    },
]


def _normalize(text):
    return re.sub(r"\s+", " ", (text or "").lower().strip())


def _match_any(text, keywords):
    return any(k in text for k in keywords)


def _course_list(limit=8):
    return list(Course.objects.order_by("sort_order", "title")[:limit])


def _search_courses(query, limit=5):
    from django.db.models import Q

    qs = Course.objects.all()
    if query:
        qs = qs.filter(Q(title__icontains=query) | Q(category__icontains=query))
    return list(qs.order_by("sort_order", "title")[:limit])


def _user_role(user):
    if not user or not user.is_authenticated:
        return "guest"
    return getattr(user, "lms_role", None) or getattr(user, "role", "student")


def _greeting(user):
    role = _user_role(user)
    if role == "admin":
        return (
            f"Salaam! I'm {BOT_NAME}, your platform assistant. "
            "I can help with the admin dashboard, users, courses, enrollments, and site navigation."
        )
    if role == "teacher":
        return (
            f"Salaam! I'm {BOT_NAME}. I can help you manage courses, understand enrollments, "
            "and guide learners on Barwaaqo Skills."
        )
    if role == "student":
        name = user.display_name if user else "there"
        return (
            f"Salaam {name}! I'm {BOT_NAME}. Ask me about your courses, XP, certificates, "
            "or where to go next on the platform."
        )
    return (
        f"Salaam! I'm {BOT_NAME}, the Barwaaqo Skills helper. "
        "I can explain our free courses, signup, and how learning works here."
    )


def _suggestions_for_role(role):
    if role == "admin":
        return [
            "Open admin dashboard",
            "Manage students",
            "View system activity",
            "Add a new course",
        ]
    if role == "teacher":
        return [
            "Browse courses",
            "How do certificates work?",
            "Enrollment limit",
            "Go to my profile",
        ]
    if role == "student":
        return [
            "Resume my course",
            "How does XP work?",
            "Browse courses",
            "Get a certificate",
        ]
    return [
        "How do I sign up?",
        "Are courses free?",
        "Browse courses",
        "What is Barwaaqo Skills?",
    ]


def _try_openai(message, user, page_path, history):
    api_key = os.getenv("OPENAI_API_KEY", "").strip()
    if not api_key:
        return None
    try:
        from openai import OpenAI

        role = _user_role(user)
        courses = ", ".join(c.title for c in _course_list(12)) or "web development, design, business"
        system = f"""You are {BOT_NAME}, the friendly AI assistant for Barwaaqo Skills — a free online learning platform.
User role: {role}. Current page: {page_path or "unknown"}.
Courses include: {courses}.
Rules: all courses are free; max 3 active enrollments; email verification required; 10 XP per lesson, 100 per course; certificates after quiz pass.
Answer clearly in the user's language if they write in Somali or Arabic, otherwise English. Keep answers concise (2-4 sentences). Be warm and helpful."""

        messages = [{"role": "system", "content": system}]
        for item in (history or [])[-6:]:
            role_key = "user" if item.get("role") == "user" else "assistant"
            content = (item.get("content") or "").strip()
            if content:
                messages.append({"role": role_key, "content": content})
        messages.append({"role": "user", "content": message})

        client = OpenAI(api_key=api_key)
        resp = client.chat.completions.create(
            model=os.getenv("OPENAI_MODEL", "gpt-4o-mini"),
            messages=messages,
            max_tokens=350,
            temperature=0.6,
        )
        text = (resp.choices[0].message.content or "").strip()
        return text if text else None
    except Exception:
        return None


def _rule_reply(message, user, page_path):
    text = _normalize(message)
    role = _user_role(user)

    if not text or text in {"hi", "hello", "hey", "salaam", "asc", "asalamu alaykum"}:
        return _greeting(user)

    if _match_any(text, ["who are you", "your name", "sharshabeel", "شرشبيل"]):
        return (
            f"I'm {BOT_NAME} — the Barwaaqo Skills assistant. "
            "I help guests, students, teachers, and admins find answers and navigate the platform."
        )

    if _match_any(text, ["barwaaqo", "what is this", "about platform", "about site"]):
        return (
            "Barwaaqo Skills is a free learning platform with career-focused courses in coding, design, "
            "business, and personal growth. Learn at your own pace and track XP on your dashboard."
        )

    if _match_any(text, ["sign up", "signup", "register", "create account", "join"]):
        return "Create a free account at signup.html, then verify your email. After that you can enroll in up to 3 courses."

    if _match_any(text, ["log in", "login", "sign in"]):
        return "Go to login.html. If you're an admin, use your admin credentials — you'll be taken to the admin dashboard automatically."

    if _match_any(text, ["admin", "lms", "dashboard admin"]) and role == "admin":
        return (
            "Your admin dashboard is at lms.html. From there you can manage students, teachers, courses, "
            "enrollments, certificates, and system activity (visits & watch history)."
        )

    if _match_any(text, ["admin", "administrator"]) and role != "admin":
        return "The admin dashboard is only for platform administrators. If you need account help, ask about courses or your profile."

    if _match_any(text, ["student", "manage user", "system activity", "visits"]) and role == "admin":
        return "In the admin panel, open Students or System activity from the sidebar to view users, page visits, and who is watching which course."

    if _match_any(text, ["profile", "photo", "avatar", "picture"]):
        if role == "admin":
            return "Open profile.html (My account in the admin sidebar). Click the camera icon on your avatar to upload a profile photo."
        if role in ("student", "teacher"):
            return "Go to profile.html to update your name and profile photo. Click the camera icon on your avatar to upload an image."
        return "After you sign up and log in, visit profile.html to manage your account and profile photo."

    if _match_any(text, ["teacher", "instructor"]):
        if role == "teacher":
            return "As a teacher, you can be assigned courses by an admin. Visit profile.html for your account, or browse courses.html to preview the catalog."
        if role == "admin":
            return "Manage teachers at lms.html#teachers — add instructors, assign roles, and link them to courses under Manage courses."
        return "Our courses are taught by expert instructors. Sign up as a learner to enroll, or contact the platform admin about teaching."

    for entry in FAQ_ENTRIES:
        if _match_any(text, entry["keys"]):
            return entry["answer"]

    if _match_any(text, ["course", "learn", "catalog", "browse", "class"]):
        cat = None
        for c in ("coding", "design", "business", "personal"):
            if c in text:
                cat = c
                break
        courses = _search_courses(cat or "", 5)
        if courses:
            titles = ", ".join(f"**{c.title}** ({c.category})" for c in courses)
            return f"Here are some courses you might like: {titles}. Open courses.html to explore and enroll."
        return "Browse all courses at courses.html. Every course is free once you're signed in and verified."

    if _match_any(text, ["resume", "continue", "where left", "last lesson"]):
        if role in ("student", "teacher", "admin"):
            return "Open dashboard.html — the Continue learning card takes you back to your last lesson. Or check My Courses for each enrollment."
        return "After you enroll, your dashboard saves your place automatically. Sign up first, then start from dashboard.html."

    if _match_any(text, ["dashboard", "my courses", "home page"]):
        if role == "admin":
            return "Admins use lms.html as the main hub. Learners use dashboard.html and my-courses.html for their progress."
        if role != "guest":
            return "Your dashboard is at dashboard.html. Active enrollments are listed on my-courses.html."
        return "After login, dashboard.html shows your progress, XP, and continue-learning shortcuts."

    if _match_any(text, ["help", "support", "faq", "stuck"]):
        return "Check faq.html for common questions, or ask me directly. I can help with signup, courses, XP, certificates, and navigation."

    # Course title search
    for course in _course_list(25):
        slug_part = course.slug.replace("-", " ")
        if course.slug in text or slug_part in text or course.title.lower() in text:
            return (
                f"**{course.title}** is in {course.category}. "
                f"{course.description[:180]}… "
                f"Preview at course-preview.html?id={course.slug}"
                + (
                    f" or enroll from courses.html."
                    if role != "guest"
                    else " — sign up free to enroll."
                )
            )

    return None


def sharshabeel_reply(message, user=None, page_path="", history=None, request=None):
    message = (message or "").strip()
    if not message:
        return _greeting(user), _suggestions_for_role(_user_role(user))

    ai = _try_openai(message, user, page_path, history)
    if ai:
        return ai, _suggestions_for_role(_user_role(user))

    reply = _rule_reply(message, user, page_path)
    if reply:
        return reply, _suggestions_for_role(_user_role(user))

    role = _user_role(user)
    fallback = (
        f"I'm not sure about that yet, but I can help with courses, signup, XP, certificates, and navigation. "
        f"Try one of the suggestions below, or visit faq.html for more."
    )
    if role == "admin":
        fallback = (
            "I can help with the admin dashboard, users, courses, enrollments, and system activity. "
            "Try asking about students, teachers, or visits."
        )
    return fallback, _suggestions_for_role(role)

File: js/lib/sharshabeel.js
/** شرشبيل — Barwaaqo Skills AI assistant */
window.BSSharshabeel = {
  BOT_NAME: "شرشبيل",
  _open: false,
  _history: [],
  _sending: false,

  _role() {
    if (!window.BSAuth?.isLoggedIn?.()) return "guest";
    if (window.BSAuth.isAdmin()) return "admin";
    if (window.BSAuth.isTeacher?.()) return "teacher";
    if (window.BSAuth.isVerified()) return "student";
    return "guest";
  },

  _pagePath() {
    const page = location.pathname.split("/").pop() || "index.html";
    return page + location.search;
  },

  _formatReply(text) {
    return (text || "")
      .replace(/\*\*(.+?)\*\*/g, "<strong>$1</strong>")
      .replace(/\n/g, "<br>");
  },

  _localFallback(message) {
    const t = message.toLowerCase();
    if (/hi|hello|hey|salaam|asc/.test(t)) {
      return "Salaam! I'm شرشبيل. I can help with courses, signup, XP, and navigation on Barwaaqo Skills.";
    }
    if (/free|cost|price/.test(t)) {
      return "All courses are free. Sign up, verify your email, and enroll in up to 3 courses at a time.";
    }
    if (/sign up|signup|register/.test(t)) {
      return "Create your free account at signup.html, then verify your email to start learning.";
    }
    if (/course|learn/.test(t)) {
      return "Browse our free catalog at courses.html — coding, design, business, and more.";
    }
    return "I'm here to help! Ask about courses, signup, XP, certificates, or where to go on the site. You can also visit faq.html.";
  },

  _defaultSuggestions() {
    const role = this._role();
    if (role === "admin") {
      return ["Open admin dashboard", "Manage students", "System activity", "Are courses free?"];
    }
    if (role === "teacher") {
      return ["Browse courses", "How do certificates work?", "Go to profile", "Enrollment limit"];
    }
    if (role === "student") {
      return ["Resume my course", "How does XP work?", "Browse courses", "Get a certificate"];
    }
    return ["How do I sign up?", "Are courses free?", "Browse courses", "What is Barwaaqo Skills?"];
  },

  _roleLabel() {
    const map = {
      guest: "Here to help visitors",
      student: "Your learning guide",
      teacher: "Instructor assistant",
      admin: "Admin platform helper",
    };
    return map[this._role()] || map.guest;
  },

  _els() {
    return {
      fab: document.getElementById("sharshabeelFab"),
      panel: document.getElementById("sharshabeelPanel"),
      messages: document.getElementById("sharshabeelMessages"),
      input: document.getElementById("sharshabeelInput"),
      send: document.getElementById("sharshabeelSend"),
      suggestions: document.getElementById("sharshabeelSuggestions"),
      subtitle: document.getElementById("sharshabeelSubtitle"),
    };
  },

  _scrollMessages() {
    const { messages } = this._els();
    if (messages) messages.scrollTop = messages.scrollHeight;
  },

  _addMessage(role, content) {
    const { messages } = this._els();
    if (!messages) return;
    const div = document.createElement("div");
    div.className = `sharshabeel-msg sharshabeel-msg--${role === "user" ? "user" : "bot"}`;
    if (role === "user") {
      div.textContent = content;
    } else {
      div.innerHTML = this._formatReply(content);
    }
    messages.appendChild(div);
    this._scrollMessages();
  },

  _setTyping(on) {
    const { messages } = this._els();
    if (!messages) return;
    const existing = messages.querySelector(".sharshabeel-typing");
    if (!on) {
      existing?.remove();
      return;
    }
    if (existing) return;
    const el = document.createElement("div");
    el.className = "sharshabeel-typing";
    el.textContent = "شرشبيل is typing…";
    messages.appendChild(el);
    this._scrollMessages();
  },

  _renderSuggestions(list) {
    const { suggestions } = this._els();
    if (!suggestions) return;
    const items = (list && list.length ? list : this._defaultSuggestions()).slice(0, 4);
    suggestions.innerHTML = items
      .map((s) => `<button type="button" class="sharshabeel-chip" data-suggest="${s.replace(/"/g, "&quot;")}">${s}</button>`)
      .join("");
    suggestions.querySelectorAll("[data-suggest]").forEach((btn) => {
      btn.addEventListener("click", () => this.send(btn.dataset.suggest));
    });
  },

  toggle(force) {
    this._open = typeof force === "boolean" ? force : !this._open;
    const { fab, panel, input } = this._els();
    fab?.classList.toggle("open", this._open);
    panel?.classList.toggle("open", this._open);
    if (this._open) {
      input?.focus();
      if (this._history.length === 0) {
        this._welcome();
      }
    }
  },

  _welcome() {
    const role = this._role();
    let text = "Salaam! I'm شرشبيل, your Barwaaqo Skills assistant. How can I help you today?";
    if (role === "admin") {
      text = "Salaam! I'm شرشبيل. I can help you with the admin dashboard, users, courses, and platform activity.";
    } else if (role === "student" && window.BSAuth?.user?.name) {
      text = `Salaam ${window.BSAuth.user.name.split(" ")[0]}! I'm شرشبيل. Ask me about your courses, XP, or certificates.`;
    }
    this._addMessage("bot", text);
    this._history.push({ role: "assistant", content: text });
    this._renderSuggestions();
  },

  async send(rawMessage) {
    const message = (rawMessage || this._els().input?.value || "").trim();
    if (!message || this._sending) return;

    const { input, send } = this._els();
    if (input) input.value = "";
    this._addMessage("user", message);
    this._history.push({ role: "user", content: message });
    this._sending = true;
    if (send) send.disabled = true;
    this._setTyping(true);

    let reply = null;
    let suggestions = null;

    try {
      if (window.BSAPI) {
        const data = await window.BSAPI.post("/learning/sharshabeel/", {
          message,
          pagePath: this._pagePath(),
          history: this._history.slice(-8),
        });
        reply = data.reply;
        suggestions = data.suggestions;
      }
    } catch {
      reply = null;
    }

    this._setTyping(false);

    if (!reply) {
      reply = this._localFallback(message);
      suggestions = this._defaultSuggestions();
    }

    this._addMessage("bot", reply);
    this._history.push({ role: "assistant", content: reply });
    this._renderSuggestions(suggestions);

    this._sending = false;
    if (send) send.disabled = false;
    input?.focus();
  },

  refresh() {
    const { subtitle } = this._els();
    if (subtitle) subtitle.textContent = this._roleLabel();
    if (this._open && this._history.length === 0) {
      this._renderSuggestions();
    }
  },

  mount() {
    if (document.getElementById("sharshabeelFab") || document.body?.dataset?.page === "design") return;

    const wrap = document.createElement("div");
    wrap.id = "sharshabeelRoot";
    wrap.innerHTML = `
      <button type="button" class="sharshabeel-fab" id="sharshabeelFab" aria-label="Open شرشبيل assistant">
        <span class="sharshabeel-fab-label">Ask شرشبيل</span>
        <i class="fa-solid fa-comments"></i>
      </button>
      <div class="sharshabeel-panel" id="sharshabeelPanel" role="dialog" aria-label="شرشبيل chat">
        <header class="sharshabeel-head">
          <div class="sharshabeel-avatar" aria-hidden="true"><i class="fa-solid fa-robot"></i></div>
          <div class="sharshabeel-head-text">
            <strong>شرشبيل</strong>
            <span id="sharshabeelSubtitle">${this._roleLabel()}</span>
          </div>
          <button type="button" class="sharshabeel-close" id="sharshabeelClose" aria-label="Close chat"><i class="fa-solid fa-xmark"></i></button>
        </header>
        <div class="sharshabeel-messages" id="sharshabeelMessages"></div>
        <div class="sharshabeel-suggestions" id="sharshabeelSuggestions"></div>
        <form class="sharshabeel-input-row" id="sharshabeelForm">
          <input type="text" class="sharshabeel-input" id="sharshabeelInput" placeholder="Ask شرشبيل anything…" maxlength="500" autocomplete="off">
          <button type="submit" class="sharshabeel-send" id="sharshabeelSend" aria-label="Send"><i class="fa-solid fa-paper-plane"></i></button>
        </form>
      </div>`;
    document.body.appendChild(wrap);

    document.getElementById("sharshabeelFab")?.addEventListener("click", () => this.toggle());
    document.getElementById("sharshabeelClose")?.addEventListener("click", () => this.toggle(false));
    document.getElementById("sharshabeelForm")?.addEventListener("submit", (e) => {
      e.preventDefault();
      this.send();
    });

    document.addEventListener("keydown", (e) => {
      if (e.key === "Escape" && this._open) this.toggle(false);
    });

    this.refresh();
  },
};

(function bootSharshabeel() {
  if (document.readyState === "loading") {
    document.addEventListener("DOMContentLoaded", () => window.BSSharshabeel.mount());
  } else {
    window.BSSharshabeel.mount();
  }
})();
